pkg/controller/storagecluster/reconcile.go

```diff
@@ -0,0 +1,97 @@
+	rookCephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	rook "github.com/rook/rook/pkg/apis/rook.io/v1alpha2"
+)
+
+// Reconcile reads that state of the cluster for a StorageCluster object and makes changes based on the state read
+// and what is in the StorageCluster.Spec
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileStorageCluster) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling StorageCluster")
+
+	// Fetch the StorageCluster instance
+	instance := &ocsv1alpha1.StorageCluster{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+	// Define a new CephCluster object
+	cephCluster := newCephCluster(instance)
+
+	// Set StorageCluster instance as the owner and controller
+	if err := controllerutil.SetControllerReference(instance, cephCluster, r.scheme); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Check if this CephCluster already exists
+	found := &rookCephv1.CephCluster{}
+	err = r.client.Get(context.TODO(), types.NamespacedName{Name: cephCluster.Name, Namespace: cephCluster.Namespace}, found)
+	if err != nil && errors.IsNotFound(err) {
+		reqLogger.Info("Creating a new CephCluster", "CephCluster.Namespace", cephCluster.Namespace, "CephCluster	.Name", cephCluster.Name)
+		err = r.client.Create(context.TODO(), cephCluster)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		// CephCluster created successfully - don't requeue
+		return reconcile.Result{}, nil
+	} else if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// CephCluster already exists - don't requeue
+	reqLogger.Info("Skip reconcile: CephCluster already exists", "CephCluster.Namespace", found.Namespace, "CephCluster.Name", found.Name)
+	return reconcile.Result{}, nil
+}
+
+// newCephCluster returns a Cephcluster object that doesn't point at any backing storage.
+func newCephCluster(sc *ocsv1alpha1.StorageCluster) *rookCephv1.CephCluster {
+	labels := map[string]string{
+		"app": sc.Name,
+	}
+	return &rookCephv1.CephCluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      sc.Name,
+			Namespace: sc.Namespace,
+			Labels:    labels,
+		},
+		Spec: rookCephv1.ClusterSpec{
+			CephVersion: rookCephv1.CephVersionSpec{
+				Image:            "ceph/ceph:v14.2.1-20190430",
```

Comment:
I believe this should not be hard-coded and should come from maybe a field in CR or an ENV variable. I wouldn't want to change code and rebuild everytime I want to update my ceph cluster.

Reply:
In this case we do. We want to control the exact images used for our containers at the operator level, with no ability to modify them, meaning that a given release of the operator guarantees a given set of container image versions. And yes, this does mean putting out a new release of the operator for every single image change.

If for some reason an admin wants to experiment beyond our prescribe configurations, they can always freely create their own CephCluster CRs by hand. :)

Reply:
ack.

Reply:
The OCS Operator will ship with a fixed image for Rook-Ceph. When the OCS Operator is updated, it will update any CephCluster CRs with the new image as appropriate. Currently there is no strict plan for how to handle upgrades, but I imagine it will look something like this:

1. Check the image name in the current CephCluster CR
1. If it is different (in any way) from the one the OCS Operator is using:
  1. Update the CephCluster CR
  1. Update the OCS Pods one at a time (probably just by deleting the OSD Pods)